otlp/traces_direct_json.go

```diff
@@ -0,0 +1,1115 @@
+				}
+				return
+			}
+
+			attrs.addFloat64(key, floatVal)
+
+		case "bytesValue":
+			// Bytes are base64 encoded in JSON
+			strBytes := v.GetStringBytes()
+			var b []byte
+			b, err = base64.StdEncoding.DecodeString(string(strBytes))
+			if err != nil {
+				return
+			}
+
+			husky.AddTelemetryAttribute(ctx, "received_bytes_attr_type", true)
+			attrs.addAny(key, addAttributeToMapAsJsonDirect(b))
+
+		case "arrayValue":
+			// For arrays, we need to parse and JSON encode
+			husky.AddTelemetryAttribute(ctx, "received_array_attr_type", true)
+			var arr []any
+			arr, err = unmarshalArrayValueJSON(ctx, v)
+			if err != nil {
+				return
+			}
+			attrs.addAny(key, addAttributeToMapAsJsonDirect(arr))
+
+		case "kvlistValue":
+			// For kvlists, handle flattening
+			husky.AddTelemetryAttributes(ctx, map[string]any{
+				"received_kvlist_attr_type": true,
+				"kvlist_max_depth":          depth,
+			})
+
+			if depth < maxDepth {
+				// Flatten the kvlist
+				if err = unmarshalKvlistValueFlattenJSON(ctx, v, key, attrs, depth+1); err != nil {
+					return
+				}
+			} else {
+				// Max depth exceeded, JSON encode the whole thing
+				var m map[string]any
+				m, err = unmarshalKvlistValueJSON(ctx, v)
+				if err != nil {
+					return
+				}
+				attrs.addAny(key, addAttributeToMapAsJsonDirect(m))
+			}
+		}
+	})
+
+	return err
+}
+
+// unmarshalArrayValueJSON parses an ArrayValue from JSON and returns []any
+// Only used when we're going to re-marshal the values as JSON.
+func unmarshalArrayValueJSON(ctx context.Context, v *fastjson.Value) ([]any, error) {
+	obj, err := v.Object()
+	if err != nil {
+		return nil, nil
+	}
+
+	var values []any
+
+	obj.Visit(func(key []byte, v *fastjson.Value) {
+		switch string(key) {
+		case "values":
+			valuesArray := v.GetArray()
+			if len(valuesArray) > 0 {
+				values = make([]any, 0, len(valuesArray))
+				for _, valObj := range valuesArray {
+					var val any
+					val, err = unmarshalAnyValueJSON(ctx, valObj)
+					if err != nil {
+						return
+					}
+					if val != nil {
+						values = append(values, val)
+					}
+				}
+			}
+		}
+	})
+
+	return values, err
+}
+
+// unmarshalAnyValueJSON parses an AnyValue from JSON and returns the value
+// Only used when we're going to re-marshal the values as JSON.
+func unmarshalAnyValueJSON(ctx context.Context, v *fastjson.Value) (any, error) {
+	obj, err := v.Object()
+	if err != nil {
+		return nil, nil
+	}
+
+	var result any
+	obj.Visit(func(key []byte, v *fastjson.Value) {
+		switch string(key) {
+		case "stringValue":
+			result = string(v.GetStringBytes())
+
+		case "boolValue":
+			result = v.GetBool()
+
+		case "intValue":
+			// JSON encodes int64 as string
+			strBytes := v.GetStringBytes()
+			var val int64
+			val, err = strconv.ParseInt(string(strBytes), 10, 64)
+			if err == nil {
+				result = val
+			}
+
+		case "doubleValue":
+			result = v.GetFloat64()
+
+		case "bytesValue":
+			// Bytes are base64 encoded in JSON
```

Comment:
Can you add a test case for "bytesValue" and "arrayValue"?

Reply:
These are covered in TestUnmarshalTraceRequestDirect_Complete, see bytes_attr and array_attr.

Reply:
I think the `bytes_attr` and `array_attr` in the [test data](https://github.com/honeycombio/husky/blob/a566cf3beca0a84ad9e3b5ce7c1ca2ce232fd46e/otlp/traces_direct_test.go#L111), doesn't cover this code because this code is only called when a byte value or array value is nested inside a array value or a key value list

Reply:
Another good catch, I'll add these to TestUnmarshalTraceRequestDirect_NestedMapAttributes as well.